state-sync/aptos-data-client/src/client.rs

```diff
@@ -377,39 +607,84 @@ impl AptosDataClient {
             }
         }
 
-        // Log the peers, periodically.
-        sample!(
-            SampleRate::Duration(Duration::from_secs(PEER_LOG_FREQ_SECS)),
-            update_connected_peer_metrics(priority_peers.len(), regular_peers.len());
-        );
-
         Ok((priority_peers, regular_peers))
     }
 
-    /// Sends a request (to an undecided peer) and decodes the response
+    /// Sends the specified storage request to a number of peers
+    /// in the network and decodes the first successful response.
     async fn send_request_and_decode<T, E>(
         &self,
         request: StorageServiceRequest,
         request_timeout_ms: u64,
     ) -> crate::error::Result<Response<T>>
     where
-        T: TryFrom<StorageServiceResponse, Error = E>,
+        T: TryFrom<StorageServiceResponse, Error = E> + Send + Sync + 'static,
         E: Into<Error>,
     {
-        // Select a peer to service the request
-        let peer = self.choose_peer_for_request(&request).map_err(|error| {
-            debug!(
-                (LogSchema::new(LogEntry::StorageServiceRequest)
-                    .event(LogEvent::PeerSelectionError)
-                    .message("Unable to select peer for request!")
-                    .error(&error))
-            );
-            error
-        })?;
+        // Select the peers to service the request
+        let peers = self.choose_peers_for_request(&request)?;
 
-        // Send the request to the peer and transform the response
-        self.send_request_to_peer_and_decode(peer, request, request_timeout_ms)
-            .await
+        // If peers is empty, return an error
+        if peers.is_empty() {
+            return Err(Error::DataIsUnavailable(format!(
+                "No peers were chosen to service the given request: {:?}",
+                request
+            )));
+        }
+
+        // Update the metrics for the number of selected peers (for the request)
+        metrics::observe_value_with_label(
+            &metrics::MULTI_FETCHES_PER_REQUEST,
+            &request.get_label(),
+            peers.len() as f64,
+        );
+
+        // Send the requests to the peers (and gather abort handles for the tasks)
+        let mut sent_requests = FuturesUnordered::new();
+        let mut abort_handles = vec![];
+        for peer in peers {
+            // Send the request to the peer
+            let aptos_data_client = self.clone();
+            let request = request.clone();
+            let sent_request = tokio::spawn(async move {
+                aptos_data_client
+                    .send_request_to_peer_and_decode(peer, request, request_timeout_ms)
+                    .await
+            });
+            let abort_handle = sent_request.abort_handle();
```

Comment:
Any idea what the behavior is on the server side for a request aborted like this?
Would it immediately stop sending the response?

Reply:
I think it depends. The `abort` essentially just kills the tokio task (i.e., stops driving the poller). So, I think it really depends on when it is killed. If it's killed before the request is sent, then the server won't see anything, but it's probably more likely that it's killed after a request has already been sent. In which case, we'll probably just drop the response in the networking stack/layer (i.e., the message will be read off the TCP stream, passed to the RPC handler and then dropped because the task holding the channel end has been dropped). Does that make sense?

Reply:
Makes sense. If forge tests are running fine, it should be fine. Maybe we want to audit the logs to see if network stack is spamming any scary logs because of this.

Reply:
Yeah! I took a quick look and I think we should be okay once the next PR lands: https://github.com/aptos-labs/aptos-core/pull/11233. This PR downgrades the `Application layer unexpectedly dropped response channel` warning log (which is already pretty bad today 😄). But, I'll definitely re-run a PFN and sanity check the logs with all the commits.